### ISSUE TITLE
kv: reduce log spam post-decommissioning of a node

### ISSUE THREAD

**Issue by tomholub** [cockroachdb/cockroach#32581]
**Describe the problem**

I keep seeing warnings about a node that was decommissioned and taken offline. The node is showing as decommissioned in Admin UI (correctly).

If the connection needs to be retried periodically, logging INFO rather than WARNING would make sense for decommissioned nodes.

```
2018-11-24 18:07:18	WARNING	
grpc: addrConn.createTransport failed to connect to {173.199.114.249:26257 0  <nil>}. Err :connection error: desc = "transport: Error while dialing dial tcp 173.199.114.249:26257: i/o timeout". Reconnecting...
vendor/google.golang.org/grpc/clientconn.go:1293

2018-11-24 18:07:18	WARNING	
grpc: addrConn.createTransport failed to connect to {173.199.114.249:26257 0  <nil>}. Err :connection error: desc = "transport: Error while dialing cannot reuse client connection". Reconnecting...
vendor/google.golang.org/grpc/clientconn.go:1293

2018-11-24 18:07:18	WARNING	
Failed to dial 173.199.114.249:26257: context canceled; please retry.
vendor/google.golang.org/grpc/clientconn.go:942
```

**To Reproduce**

1. 3 node cluster, all have each other's IPs in --join
2. Add 4th node, set repl. factor to 4, wait until synced
3. Set repl. factor to 3, shutdown unwanted node
4. Decommission unwanted node through one of the live nodes
5. Check logs on other nodes a day later

The replication factors likely don't have effect.

**Expected behavior**
Mark failed attempts to connect to _decommissioned_ nodes as INFO, or stop attempting to reconnect to decommissioned nodes, or try to connect to decommissioned nodes less frequently.

**Comment by rytaft** [comment]
cc @tbg @solongordon

**Comment by sknss** [comment]
Hi @rytaft , is there any update on this issue?

**Comment by ajwerner** [comment]
The log itself is coming from gRPC [here](https://github.com/grpc/grpc-go/blob/ed10349f451e96cf671d270545c99ff0041fea7c/clientconn.go#L1087) so controlling the verbosity will be tough. That being said, it's curious that we continue to try to connect to the node. I'll attempt to reproduce the situation and see what I see.

**Comment by ajwerner** [comment]
This actually turns out to be pretty interesting. The reason we keep attempting to connect to the decommissioned node seems to be due to a stale entry in the DistSender's RangeDescriptorCache. Each time we go to send a request to a range we pull a range descriptor from the cache and then check on the connection health for each of the replicas we've decided we might send the request to (based on the cache descriptor), see [here](https://github.com/cockroachdb/cockroach/blob/b5d526ced892f1cebcd62635775c3d530fa37231/pkg/kv/transport.go#L107). That HealthCheck attempts to dial a new connection. Eventually that health check will return an error and the DistSender will go about its business sending to the request happily to the very much alive lease holder.

It seems like we need a mechanism to flush removed replicas from the range descriptor cache.

**Comment by ajwerner** [comment]
A solution discussed with @nvanbenschoten is to add range sequence numbers to batch request headers and then have the server detect if the supplied sequence number doesn't match then return RangeInfo in the response. The DistSender then could check the sequence number in the returned RangeInfo and update its cache accordingly.

**Comment by stevenliang16** [comment]
I have also seen this warning with Cockroach v2.1.7. 
@ajwerner Is there any timeline for fixing this issue? It is adding too much noise to the logs. Thanks in advance!

**Comment by tbg** [comment]
@ajwerner in the short term, a rolling restart of the cluster would also resolve the problem, right?

**Comment by ajwerner** [comment]
Yes, a rolling restart of the cluster should resolve the problem.

**Comment by developmentalmadness** [comment]
+1 on this issue for me. 

I'd like to add that our cluster size (3) is currently equal to our replication factor (also 3). So while I usually add a node so I can perform maintenance I can't just do a rolling restart because if I add a node then when I remove after I finish I'm back to the same issue mentioned here.

**Comment by ajwerner** [comment]
Interestingly it seems like this issue is somewhat related to https://github.com/cockroachdb/cockroach/issues/44053. One source of dialing nodes which no longer exists is due to the check of `ConnHealth` in the grpcTransport. The below line will asynchronously create a connection to all nodes in the range descriptor, even if that range descriptor is stale. If we were better about evicting stale range descriptors we'd see a lot less of this noise.

https://github.com/cockroachdb/cockroach/blob/13af75f3d1376fff31d1419df2951c2d8a810efd/pkg/kv/transport.go#L101

Another option is to not create the connection on the call to ConnHealth. That's a little bit risky though. While it is possible, especially on very large clusters, that there is no connection which exists between a pair of nodes for, say, replication, this call will open a connection and then hit an error. Subsequent calls will not.

**Comment by irfansharif** [comment]
Saw this in https://github.com/cockroachdb/cockroach/issues/46037 as well. Closing that issue in favor of this.

**Comment by irfansharif** [comment]
@andreimatei, @knz: I heard y'all are thinking about flushing out internal caches and such when nodes get decommissioned. Is that right? If so, this issue may automatically get resolved by that effort.

**Comment by knz** [comment]
yes it's likely.

**Comment by knz** [comment]
This has been addressed mainly by #51937, and will be improved further if/when #53168 merges.

Many of the root causes in the orignal issue have been addressed by the range desc cache refactor, and the decommission refactor. I think this issue as-is does not need to exit any more, and if we see new symptoms we'll need to take it from there.
